### ISSUE TITLE
Shutting down server gracefully

### ISSUE THREAD

**Issue by kachayev** [clj-commons/aleph#365]
There was a similar [discussion](https://github.com/ztellman/aleph/issues/309) which is already closed, but the issue is still reproducible, i.e. [here](https://travis-ci.org/ztellman/aleph/jobs/351663601#L797).  It's easy to catch "Executor is shutdown!" in [these](https://github.com/ztellman/aleph/blob/d6cb8cbb0b29d0b6d0bea69142551d377a8e67cd/test/aleph/http_test.clj#L303-L312) test cases when closing the server after the client caught timeout **while** response callbacks are still producing answers. 

What's going on here.

1. Calling `close` on the server closes [socket channel](https://github.com/ztellman/aleph/blob/d6cb8cbb0b29d0b6d0bea69142551d377a8e67cd/src/aleph/netty.clj#L921) and [event loop group](https://github.com/ztellman/aleph/blob/d6cb8cbb0b29d0b6d0bea69142551d377a8e67cd/src/aleph/netty.clj#L922).

2. After that we call [`netty/wait-for-close`](https://github.com/ztellman/aleph/blob/d6cb8cbb0b29d0b6d0bea69142551d377a8e67cd/src/aleph/netty.clj#L927).

Note, that `.shutdownGracefully` is used which yields `terminationFuture` and returns immediately, and right after that is done `on-close` callback is [called](https://github.com/ztellman/aleph/blob/d6cb8cbb0b29d0b6d0bea69142551d377a8e67cd/src/aleph/netty.clj#L923). With the default server configuration `executor` is set to newly created `flow/untilization-executor` [here](https://github.com/ztellman/aleph/blob/d6cb8cbb0b29d0b6d0bea69142551d377a8e67cd/src/aleph/http/server.clj#L451) and `shutdown-executor?` is set to [`true`](https://github.com/ztellman/aleph/blob/d6cb8cbb0b29d0b6d0bea69142551d377a8e67cd/src/aleph/http/server.clj#L442), meaning that `on-close` [callback](https://github.com/ztellman/aleph/blob/d6cb8cbb0b29d0b6d0bea69142551d377a8e67cd/src/aleph/http/server.clj#L470-L471) will shutdown executor immediately after `.shutdownGracefully` is called on the EventLoopGroup. Which might happen **before** or **after** the group is actually terminated. The former causes the problems as a response handler might not be very happy with [executor](https://github.com/ztellman/aleph/blob/d6cb8cbb0b29d0b6d0bea69142551d377a8e67cd/src/aleph/http/server.clj#L159) being shutdown.

There's a workaround for that: set `:shutdown-executor?` to `false`, create own `ExecutorService` and pass it to the server as a parameter, - when you need to shutdown the server call `.close` first, than `netty/wait-for-close` and after that `.awaitTermination` on your executor. 

I think we need more "thoughtful" default behavior. What I suggest: 

1. Run `on-close` only after the group is terminated by blocking before termination is done (might break backward compatibility)

OR

2. Run `on-close` only after the group is terminated by setting a callback for a termination future inside `.close` (might break backward compatibility "slightly")

OR

3. To extend `AlephServer` with `await-termination` function (or `shutdown-gracefully`) to take care of finishing all running response callbacks, closing all open connections, stoping netty resources and appropriate executors.

@ztellman Let me know what you think. Happy to do a PR (we've already implemented a custom termination logic for our project pretty long time ago, so I will be happy to see first-class support in the library for this scenario).

**Comment by ztellman** [comment]
I think number 2 is the best near-term fix.  Longer-term I think we should add a graceful shutdown mechanism to Aleph which causes it to stop accepting new connections, waits for all existing connections to close, and only then closes all associated resources (with some sort of timeout that will short-circuit everything).

**Comment by kachayev** [comment]
@ztellman Number 2 is done. Logs are clean when running tests now.

Regarding longer-term solution, I think that's actually what `.shutdownGracefully` is doing in Netty. We just need to wrap it carefully. I'll dig into that a bit later.

**Comment by ztellman** [comment]
I believe `shutdownGracefully` affects the executor, but not the behavior of the server itself.  We want to preclude any future attempted use of the executor, not just flush out any remaining tasks and wait for an idle period.

**Comment by kachayev** [comment]
@ztellman Yeah, you're right. Netty implementation is very "graceful" as to say.

**Comment by malcolmsparks** [comment]
I just want to add that with aleph 0.4.5-alpha6, after a `.close` on the server, a restart with `start-server` causes the connection to open but curl will just hang (no headers returned). This doesn't feel right, reloaded workflows involve constantly stopping and starting the aleph server (resets).

See https://github.com/juxt/aleph-issue/releases/tag/evidence-365 for a simple scratch project that exhibits this behavior.

To confirm, this issue doesn't exist in 0.4.5-alpha5 so a recent change has broken something.

**Comment by kachayev** [comment]
@malcolmsparks Did you try to use `netty/wait-for-close` to wait until it's actually stopped? Does it solve the issue or the problem remains?

**Comment by malcolmsparks** [comment]
When I call `.close` I get:
```
Apr 03, 2018 11:11:48 PM io.netty.util.concurrent.DefaultPromise notifyListener0
WARNING: An exception was thrown by aleph.netty$wrap_future$reify__10559.operationComplete()
java.lang.ExceptionInInitializerError
	at manifold.deferred.Deferred.success(deferred.clj:395)
	at manifold.deferred$success_BANG_.invokeStatic(deferred.clj:243)
	at manifold.deferred$success_BANG_.invoke(deferred.clj:240)
	at aleph.netty$wrap_future$reify__10559.operationComplete(netty.clj:199)
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:511)
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:485)
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:424)
	at io.netty.util.concurrent.DefaultPromise.setSuccess(DefaultPromise.java:94)
	at io.netty.util.concurrent.MultithreadEventExecutorGroup$1.operationComplete(MultithreadEventExecutorGroup.java:117)
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:511)
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:485)
	at io.netty.util.concurrent.DefaultPromise.access$000(DefaultPromise.java:33)
	at io.netty.util.concurrent.DefaultPromise$1.run(DefaultPromise.java:435)
	at io.netty.util.concurrent.GlobalEventExecutor$TaskRunner.run(GlobalEventExecutor.java:248)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassNotFoundException: clojure/lang/PersistentArrayMap
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at clojure.lang.RT.classForName(RT.java:2204)
	at clojure.lang.RT.classForName(RT.java:2213)
	at clojure.lang.Reflector.invokeStaticMethod(Reflector.java:199)
	at clojure.lang.LispReader$EvalReader.invoke(LispReader.java:1308)
	at clojure.lang.LispReader$DispatchReader.invoke(LispReader.java:843)
	at clojure.lang.LispReader.read(LispReader.java:275)
	at clojure.lang.LispReader.read(LispReader.java:210)
	at clojure.lang.LispReader.access$200(LispReader.java:41)
	at clojure.lang.LispReader$MetaReader.invoke(LispReader.java:944)
	at clojure.lang.LispReader.read(LispReader.java:275)
	at clojure.lang.LispReader.readDelimitedList(LispReader.java:1384)
	at clojure.lang.LispReader$ListReader.invoke(LispReader.java:1232)
	at clojure.lang.LispReader.read(LispReader.java:275)
	at clojure.lang.LispReader.read(LispReader.java:210)
	at clojure.lang.LispReader.access$200(LispReader.java:41)
	at clojure.lang.LispReader$EvalReader.invoke(LispReader.java:1287)
	at clojure.lang.LispReader$DispatchReader.invoke(LispReader.java:843)
	at clojure.lang.LispReader.read(LispReader.java:275)
	at clojure.lang.LispReader.read(LispReader.java:206)
	at clojure.lang.LispReader.read(LispReader.java:195)
	at clojure.lang.RT.readString(RT.java:1871)
	at clojure.lang.RT.readString(RT.java:1866)
	at manifold.deferred.Deferred$fn__6459.<clinit>(deferred.clj:395)
	... 16 more
```
It's odd. Sometimes `netty/wait-for-close` allows me to restart the server, most times I still get the 'hanging curl' issue:

```
[malcolm@melkor aleph-issue]$ curl -v localhost:8080
* Rebuilt URL to: localhost:8080/
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> GET / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.59.0
> Accept: */*
> 
```

**Comment by kachayev** [comment]
@malcolmsparks I was actually able to replicate the problem. I'm getting another `ClassNotFoundException` exception from `io.netty.util.concurrent.DefaultPromise`:

```
java.lang.NoClassDefFoundError: Could not initialize class manifold.deferred.Deferred$fn__2582
```

which is super weird. I'll try to dive deeper and find what's causing this behavior today. But, in fact,  any issue with `on-close` callback should not mess up any successive server launches. Everything that might happen: we didn't clean up an executor properly. And in the default configuration, this particular executor has nothing to do with the next launch as we will setup a new one.

**Comment by kachayev** [comment]
@ztellman I've spent quite some time trying to find the root of the problem. With no luck. It seems like something tricky from the deep inside of `manifold`. Adding listener directly to Netty's future works perfectly fine: https://github.com/ztellman/aleph/compare/master...kachayev:fix-exception-on-close

**Comment by scramjet** [comment]
Given the error occurs in the fn's class initialiser, and involves a diversion into the Clojure reader, my suspicion would be that it's something the compiler is generating (I know, I know, it's never the compiler!)

If I get some time tomorrow, I'll look at the Java output using https://github.com/clojure-goes-fast/clj-java-decompiler, which I've had some luck with in the past.

**Comment by kachayev** [comment]
I actually was able to find where the error occurs: under some circumstances

```(.success deferred x)```

forces compiler to re-eval `manifold.deferred` namespace. Meaning at that point compiler *somehow* doesn't know that namespaces are already loaded and *somehow* is not able to load `clojure.lang.*` classes. So, by default, I would say that's the compiler + RT + classloader issue.
  
<img width="1413" alt="aleph_3" src="https://user-images.githubusercontent.com/485647/47845229-c13b0800-ddc4-11e8-9e64-3a0f615f4709.png">

**Comment by kachayev** [comment]
@scramjet I'm not sure that using Clojure JIRA is a good idea *before* we have on our hands 1) minimal steps to reproduce, 2) evidence that either the compiler or runtime is causing troubles, 3) patch.

**Comment by scramjet** [comment]
@kachayev You're probably right that raising an issue isn't the right approach. Perhaps Clojurians Slack or IRC then?

**Comment by scramjet** [comment]
I just hit this again in a different context: wrapping a core.async channel in a `Deferred` for use in Aleph like so:

```
(-> (async/go
        (let [response (<! ch)]
          (hiccup->html response)))
      (manifold.stream/->source)
      (manifold.stream/take!))
```

The value from the channel never gets delivered, and I see this dropped into stderr:

```
Exception in thread "async-dispatch-32" 
java.lang.NoClassDefFoundError: Could not initialize class manifold.deferred.Deferred$fn__34063
    at manifold.deferred.Deferred.success(deferred.clj:398)
    at manifold.deferred$success_BANG_.invokeStatic(deferred.clj:243)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:240)
    at manifold.stream.async.CoreAsyncSource$f__44709$fn__44710.invoke(async.clj:44)
    at clojure.core.async.impl.channels.ManyToManyChannel$fn__36451$fn__36452.invoke(channels.clj:95)
    at clojure.lang.AFn.run(AFn.java:22)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
    at java.lang.Thread.run(Thread.java:748)
```

Keeping Aleph at 0.4.7-alpha3 and manually rolling Manifold back to 0.1.8 fixes it.

**Comment by kachayev** [comment]
@scramjet That's super weird... I don't see any related changes neither since Aleph 0.4.7-alpha3 nor since Manifold 0.1.8 🤔

**Comment by ztellman** [comment]
Sorry I'm late to the conversation.  Can you confirm that you've called `'lein clean` or some equivalent before starting the process where you see this error?  Sometimes old versions of code can cause these sorts of issues.

**Comment by kachayev** [comment]
@ztellman Yep. The issue is still reproducible even after doing `lein clean`.

**Comment by scramjet** [comment]
@kachayev I swear I blew away my `target` directory and retried before reporting the last issue. But if I do `lein clean` and roll back to Manifold `0.1.9-alpha2`, I get no errors. So I can only assume it was some issue with the way lein works that I don't understand. Will use `clean` in future. 

Sorry to throw a red herring into an already perplexing problem.

**Comment by kachayev** [comment]
@scramjet @ztellman I think I actually found the true root cause of the issue. It's kinda a long story into `clojure.lang.RT` internals but in short words, the general idea is the following: `d/success!` within the context of `wrap-future` runs on a different (netty) thread that where RT fails to find a proper class loader [here](https://github.com/clojure/clojure/blob/a1722cf9a15d917ee78a117732cf5c5e01fe3557/src/jvm/clojure/lang/RT.java#L2173-L2179). Thread local variable `LOADER` is not set and the thread has no context loader. So, it falls back to a native `Class.forName` which has no clue where to find `clojure.lang.PersistentArrayMap`. I still need to look into the bytecode of compiled `Deferred$` fns to understand why the loader is not set, but right now the simplest solution would be to switch off `*use-context-classloader*`. You can check that wrapping `d/success!` like that:

```clojure
(binding [*use-context-classloader* false]
  (d/success! d (.getNow f)))
```

makes everything to run smoothly. Will continue digging tomorrow.

**Comment by kachayev** [comment]
@ztellman Okay. So... Here we have a [try-catch block](https://github.com/ztellman/manifold/blob/d67a8c1b9f1268c094895d70dbbf146521f5774b/src/manifold/deferred.clj#L281-L293) that compiles to a separate class `Deferred$fn__<whatever>` to be lately instanciated (and invoked) in `Deferred.success`. The class contains static variable with `RT.readString` of the ns declaration, I assume that comes from a `clojure.tools.logging` toolchain.

The problem is that our first attempt to use this class happens on a netty thread (`globalEventExecutor*` to be more specific) and, as I wrote before, there's no appropriate class loader there. That happens because we initialize module in one thread to use  "not-very-compiled" code on another. Which is definitely a compiler issue @scramjet ;) I'm going to spend some time to make a short/simple reproducible example of that to track in Clojure dev JIRA. But for now... We have a few ways to deal with this:

* do not rely on the same thread delivery mechanism 
* manually make sure that loader is set (using `setContextClassloader` for the thread we're running on)
* switch off attempts to use context classloader as I mentioned before

Any thoughts?

**Comment by ztellman** [comment]
Setting the class loader [here](https://github.com/ztellman/manifold/blob/master/src/manifold/executor.clj#L43) seems to be the simplest answer, but I'm still confused on why this is such a low-probability error.  What is the normal order of operations that avoids this issue, and what's the unusual ordering that causes it?

**Comment by ztellman** [comment]
We'll also have to update the uses of `DefaultThreadFactory` like [this one](https://github.com/ztellman/aleph/blob/master/src/aleph/netty.clj#L916) to use the Manifold thread factory, which we should have done already.

**Comment by kachayev** [comment]
@ztellman Setting any loader [here](https://github.com/ztellman/manifold/blob/master/src/manifold/executor.clj#L43) would not work as we're talking about code that's running on netty's thread, no the one that's created by our thread factory. I'm working on a pull request to manifold to wrap a pretty specific [try..catch](https://github.com/ztellman/manifold/blob/d67a8c1b9f1268c094895d70dbbf146521f5774b/src/manifold/deferred.clj#L281-L293) block where we see the error.

Well... that's a pretty tricky situation I would say (quite of a "low-probability"). We have 1. static class field that invokes a reader, 2. that invokes macro (because of the logger), 3. that requires a dynamically loaded class to be expanded (because of the meta), 4. that is compiled into a separate function (because of try-catch), 5. that was not instantiated earlier (either because of being a part of a macro (?) or because of if branch (?) or both (?)), 6. that's launching from a thread where loader is not set (because of netty's executor that invokes reified listener instance). Remove one of that, and we would never see the problem.

**Comment by kachayev** [comment]
@ztellman No. At least I don't see how. What we need to "fix", is the behavior of the RT in the thread, created by this [factory](https://github.com/netty/netty/blob/00afb19d7a37de21b35ce4f6cb3fa7f74809f2ab/common/src/main/java/io/netty/util/concurrent/GlobalEventExecutor.java#L58-L59).

**Comment by kachayev** [comment]
@ztellman I prepared separate PR with thread factories being switched to use `manifold`. But it relies on `0.1.9-SNAPSHOT` being release (right now the last release on Clojars is still `0.1.9-alpha2`.

**Comment by ztellman** [comment]
I've pushed `0.1.9-SNAPSHOT`.

**Comment by kwladyka** [comment]
```clojure
(defn handler [req]
  {:status 200
   :headers {"content-type" "text/plain"}
   :body "hello!"})

(def foo (http/start-server handler {:port 8080}))

(.close foo)
```

```
Jan 11, 2019 1:33:05 PM io.netty.util.concurrent.DefaultPromise notifyListener0
WARNING: An exception was thrown by aleph.netty$wrap_future$reify__16507.operationComplete()
java.lang.NoClassDefFoundError: Could not initialize class manifold.deferred.Deferred$fn__3629
    at manifold.deferred.Deferred.success(deferred.clj:398)
    at manifold.deferred$success_BANG_.invokeStatic(deferred.clj:243)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:240)
    at aleph.netty$wrap_future$reify__16507.operationComplete(netty.clj:199)
    at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:511)
    at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:485)
    at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:424)
    at io.netty.util.concurrent.DefaultPromise.setSuccess(DefaultPromise.java:94)
    at io.netty.util.concurrent.MultithreadEventExecutorGroup$1.operationComplete(MultithreadEventExecutorGroup.java:117)
    at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:511)
    at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:485)
    at io.netty.util.concurrent.DefaultPromise.access$000(DefaultPromise.java:33)
    at io.netty.util.concurrent.DefaultPromise$1.run(DefaultPromise.java:435)
    at io.netty.util.concurrent.GlobalEventExecutor$TaskRunner.run(GlobalEventExecutor.java:248)
    at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
    at java.base/java.lang.Thread.run(Thread.java:834)
```

```
java --version
java 11.0.1 2018-10-16 LTS
```

```
{org.clojure/clojure {:mvn/version "1.10.0"}
 yada {:mvn/version "1.2.16"}
 aleph {:mvn/version "0.4.6"}
 bidi {:mvn/version "2.1.5"}
 integrant {:mvn/version "0.7.0"}}
```

**Comment by zhming0** [comment]
For anyone arrived from search engine: In my local, `aleph 0.4.7-alpha5` seems to address this issue.

**Comment by arnaudgeiser** [comment]
Probably fixed by : https://github.com/clj-commons/aleph/pull/604
